// jibo/lib/debug/sound/ChainBuilder.js
"use strict";
/**
 * @class ChainBuilder
 * @private
 * @param {AudioContext} audioContext The audio context.
 */
var ChainBuilder = (function () {
    function ChainBuilder(_context) {
        this._context = _context;
        this._firstNode = null;
        this._lastNode = null;
        this._nodes = {};
    }
    /**
     * Cleans up.
     * @method ChainBuilder#destroy
     */
    ChainBuilder.prototype.destroy = function () {
        this._nodes = null;
        this._context = null;
        this._firstNode = null;
        this._lastNode = null;
    };
    /**
     * Gets the nodes.
     * @method ChainBuilder#nodes
     * @return {Object}
     */
    ChainBuilder.prototype.nodes = function () {
        return this._nodes;
    };
    /**
     * Gets the first node.
     * @method ChainBuilder#first
     * @return {Object}
     */
    ChainBuilder.prototype.first = function () {
        return this._firstNode;
    };
    /**
     * Gets the last node.
     * @method ChainBuilder#last
     * @return {Object}
     */
    ChainBuilder.prototype.last = function () {
        return this._lastNode;
    };
    /**
     * Adds a node to the chain.
     * @method ChainBuilder#_addNode
     * @private
     * @param {*} node
     * @param {*} properties
     */
    ChainBuilder.prototype._addNode = function (node, properties) {
        // update this._bufferSourceDst - needed for .cloneBufferSource()
        var lastIsBufferSource = this._lastNode && ('playbackRate' in this._lastNode);
        if (lastIsBufferSource) {
            this._bufferSourceDst = node;
        }
        // connect this._lastNode to node if suitable
        if (this._lastNode) {
            this._lastNode.connect(node);
        }
        // update this._firstNode && this._lastNode
        if (!this._firstNode) {
            this._firstNode = node;
        }
        this._lastNode = node;
        // apply properties to the node
        for (var property in properties) {
            node[property] = properties[property];
        }
        // for chained API
        return this;
    };
    /**
     * Clones the bufferSource. Used just before playing a sound.
     * @method ChainBuilder#cloneBufferSource
     * @returns {AudioBufferSourceNode} The clone AudioBufferSourceNode.
     */
    ChainBuilder.prototype.cloneBufferSource = function () {
        console.assert(this._nodes.bufferSource, "No buffersource presents. Add one.");
        var orig = this._nodes.bufferSource;
        var clone = this._context.createBufferSource();
        clone.buffer = orig.buffer;
        clone.playbackRate.value = orig.playbackRate.value;
        clone.loop = orig.loop;
        clone.connect(this._bufferSourceDst);
        return clone;
    };
    /**
     * Adds a bufferSource.
     * @method ChainBuilder#bufferSrouce
     * @param {Object} [properties] Properties to set in the created node.
     */
    ChainBuilder.prototype.bufferSource = function (properties) {
        var node = this._context.createBufferSource();
        this._nodes.bufferSource = node;
        return this._addNode(node, properties);
    };
    /**
     * Adds a createMediaStreamSource.
     * @method ChainBuilder#mediaStreamSource
     * @param {Object} [properties] Properties to set in the created node.
     */
    // mediaStreamSource(stream:any, properties?:any):ChainBuilder
    // {
    //     const node = this._context.createMediaStreamSource(stream);
    //     this._nodes.bufferSource = node;
    //     return this._addNode(node, properties);
    // }
    /**
     * Adds a createMediaElementSource.
     * @method ChainBuilder#mediaElementSource
     * @param  {HTMLElement} element    The element to add.
     * @param {Object} [properties] Properties to set in the created node.
     */
    // mediaElementSource(element:HTMLElement, properties?:any):ChainBuilder
    // {
    //     console.assert(element instanceof HTMLAudioElement || element instanceof HTMLVideoElement);
    //     const node = this._context.createMediaElementSource(element);
    //     this._nodes.bufferSource = node;
    //     return this._addNode(node, properties);
    // }
    /**
     * Adds a panner.
     * @method ChainBuilder#panner
     * @param {Object} [properties] Properties to set in the created node.
     */
    ChainBuilder.prototype.panner = function (properties) {
        var node = this._context.createStereoPanner();
        this._nodes.panner = node;
        return this._addNode(node, properties);
    };
    /**
     * Adds an analyser.
     * @method ChainBuilder#analyser
     * @param {Object} [properties] Properties to set in the created node.
     */
    ChainBuilder.prototype.analyser = function (properties) {
        var node = this._context.createAnalyser();
        this._nodes.analyser = node;
        return this._addNode(node, properties);
    };
    /**
     * Adds a gainNode.
     * @method ChainBuilder#gainNode
     * @param {Object} [properties] Properties to set in the created node.
     */
    ChainBuilder.prototype.gainNode = function (properties) {
        var node = this._context.createGain();
        this._nodes.gainNode = node;
        return this._addNode(node, properties);
    };
    return ChainBuilder;
}());
Object.defineProperty(exports, "__esModule", { value: true });
exports.default = ChainBuilder;
