// jibo/lib/debug/bt/behaviors/ExecuteScriptAsync.js
"use strict";
var __extends = (this && this.__extends) || function (d, b) {
    for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p];
    function __() { this.constructor = d; }
    d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
};
var Behavior_1 = require('../Behavior');
var Status_1 = require('../Status');
/**
 * @callback module:jibo/bt/behaviors.ExecuteScriptAsync~ExecuteFunction
 * @description This behavior calls this function. ExecuteScriptAsync will succeed or fail only when one of the
 * two callbacks are called.
 * @param {Function} succeed Call this function when you want this behavior to succeed.
 * @param {Function} fail Call this function when you want this behavior to fail.
 */
/**
 * @class ExecuteScriptAsync
 * @extends module:jibo/bt.Behavior
 * @memberof module:jibo/bt/behaviors
 * @description Asynchronously executes JavaScript.
 * @param {Object} options See {@link module:jibo/bt.Behavior|Behavior} for all options.
 * @param {module:jibo/bt/behaviors.ExecuteScriptAsync~ExecuteFunction} options.exec Function to execute. Behavior will not stop unless one of the callbacks
 * 												 are called.
 */
var ExecuteScriptAsync = (function (_super) {
    __extends(ExecuteScriptAsync, _super);
    function ExecuteScriptAsync(options) {
        _super.call(this, options);
        this.status = Status_1.default.INVALID;
    }
    ExecuteScriptAsync.prototype.start = function () {
        this.status = Status_1.default.IN_PROGRESS;
        this.options.exec(this.onSucceed.bind(this), this.onFail.bind(this));
        return true;
    };
    ExecuteScriptAsync.prototype.onSucceed = function () {
        this.status = Status_1.default.SUCCEEDED;
    };
    ExecuteScriptAsync.prototype.onFail = function () {
        this.status = Status_1.default.FAILED;
    };
    ExecuteScriptAsync.prototype.update = function () {
        return this.status;
    };
    return ExecuteScriptAsync;
}(Behavior_1.default));
Object.defineProperty(exports, "__esModule", { value: true });
exports.default = ExecuteScriptAsync;
