// jibo/lib/debug/bt/behaviors/ListenJs.js
"use strict";
var __extends = (this && this.__extends) || function (d, b) {
    for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p];
    function __() { this.constructor = d; }
    d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
};
var Behavior_1 = require('../Behavior');
var SucceedOnListenJs_1 = require('../decorators/SucceedOnListenJs');
var Null_1 = require('./Null');
/**
 * @typedef {Object} module:jibo/bt/behaviors.ListenJs~Options
 * @property {boolean} heyJibo Listen for "Hey, Jibo" first.
 * @property {boolean} detectEnd Listen for end of speech.
 * @property {boolean} incremental Return incremental ASR results as they are streamed from the cloud.
 * @property {String} authenticateSpeaker Authenticates against that person.
 */
/**
 * Function that returns the options object.
 * @callback module:jibo/bt/behaviors.ListenJs~GetOptions
 * @returns {module:jibo/bt/behaviors.ListenJs~Options}
 */
/**
 * Called when the {@link module:jibo/bt/behaviors.ListenEmitter|ListenEmitter} is constructed.
 * @callback module:jibo/bt/behaviors.ListenJs~OnResult
 * @param {module:jibo/bt.ListenEmitter} listener Use this instance to listen for listen events.
 */
/**
 * @class ListenJs
 * @extends module:jibo/bt.Behavior
 * @memberof module:jibo/bt/behaviors
 * @description Performs audio speech recognition and applies and parses the results according to a rules file.
 * @param {Object} options See {@link module:jibo/bt.Behavior|Behavior} for all options.
 * @param {module:jibo/bt/behaviors.ListenJs~GetOptions} options.getOptions Returns the options object.
 * @param {Function} options.getRule This function returns a string representation of a rule. Use this behavior to dynamically generate rules files instead
 *                      of loading a rule file from disk.
 * @param {module:jibo/bt/behaviors.ListenJs~OnResult} options.onResult Called and passed a {@link module:jibo/bt/behaviors.ListenEmitter|ListenEmitter} object. Events are fired from the Listener
 *                                   at certain points in this behavior's lifecycle.
 */
var ListenJs = (function (_super) {
    __extends(ListenJs, _super);
    function ListenJs(options) {
        _super.call(this, options);
        var decorator = new SucceedOnListenJs_1.default({
            getOptions: options.getOptions,
            getRule: options.getRule,
            onResult: options.onResult,
            emitter: this.emitter,
            blackboard: this.blackboard,
            assetPack: this.assetPack
        });
        this.behavior = new Null_1.default({
            decorators: [decorator]
        });
    }
    ListenJs.prototype.start = function () {
        return this.behavior._start();
    };
    ListenJs.prototype.stop = function () {
        this.behavior._stop();
    };
    ListenJs.prototype.update = function () {
        return this.behavior._update();
    };
    return ListenJs;
}(Behavior_1.default));
Object.defineProperty(exports, "__esModule", { value: true });
exports.default = ListenJs;
