// jibo-keyframes/lib/debug/layers/EyeLayer.js
'use strict';

Object.defineProperty(exports, "__esModule", {
    value: true
});

var _createClass = function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; }();

var _EyeAndOverlayLayer2 = require('./EyeAndOverlayLayer');

var _EyeAndOverlayLayer3 = _interopRequireDefault(_EyeAndOverlayLayer2);

var _Conversion = require('../utils/Conversion');

var _Conversion2 = _interopRequireDefault(_Conversion);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _possibleConstructorReturn(self, call) { if (!self) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return call && (typeof call === "object" || typeof call === "function") ? call : self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; }

var EyeInfo = _EyeAndOverlayLayer3.default.extend({
    layerType: "Eye",
    dofs: {
        eyeSubRootBn_r: true,
        eyeSubRootBn_t: true,
        eyeSubRootBn_t_2: true,
        vertexJoint1_t: true,
        vertexJoint1_t_2: true,
        vertexJoint2_t: true,
        vertexJoint2_t_2: true,
        vertexJoint3_t: true,
        vertexJoint3_t_2: true,
        vertexJoint4_t: true,
        vertexJoint4_t_2: true,
        vertexJoint5_t: true,
        vertexJoint5_t_2: true,
        vertexJoint6_t: true,
        vertexJoint6_t_2: true,
        vertexJoint7_t: true,
        vertexJoint7_t_2: true,
        vertexJoint8_t: true,
        vertexJoint8_t_2: true,
        vertexJoint9_t: true,
        eyeVisibilityBn_r: true,
        eye_alphaChannelBn_r: true
    }
});

var EyeLayer = function (_EyeAndOverlayLayer) {
    _inherits(EyeLayer, _EyeAndOverlayLayer);

    function EyeLayer() {
        _classCallCheck(this, EyeLayer);

        return _possibleConstructorReturn(this, Object.getPrototypeOf(EyeLayer).apply(this, arguments));
    }

    _createClass(EyeLayer, null, [{
        key: 'getInfo',
        value: function getInfo() {
            return EyeInfo;
        }
    }, {
        key: 'generateDofs',
        value: function generateDofs(props) {
            var dofs = _EyeAndOverlayLayer3.default._generateDofs(props, "");
            dofs.eyeSubRootBn_r = props.Rotate / 180 * Math.PI;
            dofs.eyeSubRootBn_t = _Conversion2.default.toMetersX(props.Translate.x);
            dofs.eyeSubRootBn_t_2 = _Conversion2.default.toMetersY(props.Translate.y);
            dofs.eyeVisibilityBn_r = props.Visible === undefined ? 1 : props.Visible ? 1 : 0;
            dofs.eye_alphaChannelBn_r = props.Alpha === undefined ? 1 : props.Alpha;
            return dofs;
        }
    }]);

    return EyeLayer;
}(_EyeAndOverlayLayer3.default);

exports.default = EyeLayer;
