// jibo/lib/debug/jibo.js
// Only include PixiAnimate if it's not already installed
// this will be used by the rendering system
"use strict";
if (!GLOBAL.PIXI || !GLOBAL.PIXI.animate) {
    require('pixi-animate/dist/pixi-animate');
}
// Import the classes
var bt_1 = require('./bt');
var services_1 = require('./services');
var utils_1 = require('./utils');
var sound_1 = require('./sound');
var Runtime_1 = require('./Runtime');
var rendering_1 = require('./rendering');
// Import deprecated functionality
var deprecation_1 = require('./deprecation');
// Import external dependencies
var WebSocket = require('ws-without-native');
// Replace libchromium's WebSocket with a node.js oriented
// polyfill to avoid any throttling from browser logic
GLOBAL.WebSocket = WebSocket;
// Polyfill Object.assign for Electron 0.29.2 and before
if (!Object.assign)
    Object.assign = require('object-assign');
if (!GLOBAL.jibo) {
    // Create the class definitions
    // and hang it on a global namespace
    // object called jibo
    var jibo = {
        bt: bt_1.default,
        services: services_1.default,
        utils: utils_1.default,
        sound: sound_1.default,
        Runtime: Runtime_1.default,
        rendering: rendering_1.default
    };
    // Deprecate before create jibo
    deprecation_1.default(jibo);
    // Export the jibo singleton
    GLOBAL.jibo = new Runtime_1.default(jibo);
}
Object.defineProperty(exports, "__esModule", { value: true });
exports.default = GLOBAL.jibo;
