// animation-utilities/lib/debug/animation-animate/timeline/LayerState.js
/**
 * @author mattb
 */

"use strict";

/**
 * @param {Time} time
 * @param {Pose} pose
 * @constructor
 */

var LayerState = function LayerState(time, pose) {
	/**
  * @return {Time}
  */
	this.getTime = function () {
		return time;
	};

	/**
  * @param {Time} newTime
  */
	this.setTime = function (newTime) {
		time = newTime;
	};

	/**
  * @return {Pose}
  */
	this.getPose = function () {
		return pose;
	};

	/**
  * @return {Array.<string>}
  */
	this.getDOFNames = function () {
		return pose.getDOFNames();
	};

	/**
  * @param {string} dofName
  * @return {number[]}
  */
	this.getDOFState = function (dofName) {
		return pose.get(dofName);
	};

	/**
  * @param {string} dofName
  * @param {number[]} dofState
  */
	this.setDOFState = function (dofName, dofState) {
		pose.set(dofName, dofState);
	};

	/**
  * @param {Time} [newTime]
  * @return {LayerState}
  */
	this.getCopy = function (newTime) {
		if (newTime === null || newTime === undefined) {
			newTime = time;
		}
		return new LayerState(newTime, pose.getCopy());
	};
};

module.exports = LayerState;
