// animation-utilities/lib/debug/geometry-info/JiboConfig.js
/**
 * @author mattb
 */

"use strict";

/**
 * @param {string} [baseGeometryURL] - base geometry config directory
 * @param {string} [robotVersion] - robot version identifier
 * @constructor
 */

var JiboConfig = function JiboConfig(baseGeometryURL, robotVersion) {
	var baseAssetURL = baseGeometryURL;
	if (baseAssetURL === undefined) {
		var findRoot = "find-root";
		findRoot = require(findRoot);
		baseAssetURL = findRoot(__dirname) + "/res/geometry-config/";
	}
	var robot = robotVersion !== undefined ? robotVersion : "P1.0";
	var robotURL = baseAssetURL + robot + "/";

	var bodyGeometryURL = robotURL + "jibo_body.geom";
	var bodySkeletonURL = robotURL + "jibo_body.skel";
	var bodyKinematicsURL = robotURL + "jibo_body.kin";

	var fullGeometryURL = robotURL + "jibo_joined.geom";
	var fullSkeletonURL = robotURL + "jibo_joined.skel";
	var fullKinematicsURL = robotURL + "jibo_joined.kin";

	var eyeGeometryURL = robotURL + "jibo_eye.geom";
	var eyeSkeletonURL = robotURL + "jibo_eye.skel";
	var eyeKinematicsURL = robotURL + "jibo_eye.kin";

	var sceneInfoURL = robotURL + "jibo.jscene";

	var dofGroupsURL = robotURL + "jibo.dofgroups";

	var limitsURL = robotURL + "jibo.lim";

	var defaultNormalURL = robotURL + "defaultNormalMap.png";

	/**
  * @return {string}
  */
	this.getRobotURL = function () {
		return robotURL;
	};

	/**
  * @return {string}
  */
	this.getBodyGeometryURL = function () {
		return bodyGeometryURL;
	};
	/**
  * @return {string}
  */
	this.getBodySkeletonURL = function () {
		return bodySkeletonURL;
	};
	/**
  * @return {string}
  */
	this.getBodyKinematicsURL = function () {
		return bodyKinematicsURL;
	};

	/**
  * @return {string}
  */
	this.getFullGeometryURL = function () {
		return fullGeometryURL;
	};
	/**
  * @return {string}
  */
	this.getFullSkeletonURL = function () {
		return fullSkeletonURL;
	};
	/**
  * @return {string}
  */
	this.getFullKinematicsURL = function () {
		return fullKinematicsURL;
	};

	/**
  * @return {string}
  */
	this.getEyeGeometryURL = function () {
		return eyeGeometryURL;
	};
	/**
  * @return {string}
  */
	this.getEyeSkeletonURL = function () {
		return eyeSkeletonURL;
	};
	/**
  * @return {string}
  */
	this.getEyeKinematicsURL = function () {
		return eyeKinematicsURL;
	};

	/**
  * @return {string}
  */
	this.getSceneInfoURL = function () {
		return sceneInfoURL;
	};

	/**
  * @return {string}
  */
	this.getDOFGroupsURL = function () {
		return dofGroupsURL;
	};

	/**
  * @return {string}
  */
	this.getLimitsURL = function () {
		return limitsURL;
	};

	/**
  * @return {string}
  */
	this.getDefaultNormalMap = function () {
		return defaultNormalURL;
	};
};

module.exports = JiboConfig;
