// jibo/lib/debug/bt/Status.js
"use strict";
/**
 * Enumeration of behavior states
 * @class Status
 * @memberof module:jibo/bt
 */
var Status;
(function (Status) {
    /**
     * The behavior succeeded.
     * @name module:jibo/bt.Status.SUCCEEDED
     * @type {String}
     * @readOnly
     * @default "SUCCEEDED"
     */
    Status[Status["SUCCEEDED"] = 0] = "SUCCEEDED";
    /**
     * The behavior failed.
     * @name module:jibo/bt.Status.FAILED
     * @type {String}
     * @readOnly
     * @default "FAILED"
     */
    Status[Status["FAILED"] = 1] = "FAILED";
    /**
     * The behavior was interrupted.
     * @name module:jibo/bt.Status.INTERRUPTED
     * @type {String}
     * @readOnly
     * @default "INTERRUPTED"
     */
    Status[Status["INTERRUPTED"] = 2] = "INTERRUPTED";
    /**
     * The behavior is in progress and hasn't returned.
     * @name module:jibo/bt.Status.IN_PROGRESS
     * @type {String}
     * @readOnly
     * @default "IN_PROGRESS"
     */
    Status[Status["IN_PROGRESS"] = 3] = "IN_PROGRESS";
    /**
     * The behavior is in an invalid state.
     * @name module:jibo/bt.Status.INVALID
     * @type {String}
     * @readOnly
     * @default "INVALID"
     */
    Status[Status["INVALID"] = 4] = "INVALID";
    /**
     * Don't update the state.
     * @name module:jibo/bt.Status.PAUSED
     * @type {String}
     * @readOnly
     * @default "PAUSED"
     */
    Status[Status["PAUSED"] = 5] = "PAUSED";
    /**
     * Only used by decorators.
     * @name module:jibo/bt.Status.WAIT
     * @type {String}
     * @readOnly
     * @default "WAIT"
     */
    Status[Status["WAIT"] = 6] = "WAIT";
})(Status || (Status = {}));
Object.defineProperty(exports, "__esModule", { value: true });
exports.default = Status;
