// animation-utilities/lib/debug/ifr-geometry/loaders/CachedImageLoader.js
/**
 * @author mattb
 * Copyright 2014 IF Robots LLC
 */

"use strict";

var THREE = require("three");

/**
 * @constructor
 */
var ImageLoadResult = function ImageLoadResult() {
	/** @type {string} */
	this.url = null;
	/** @type {!boolean} */
	this.success = false;
	/** @type {THREE.Object3D} */
	this.image = null;
};

/**
 * @constructor
 */
var CachedImageLoader = function CachedImageLoader() {
	/** @type {THREE.ImageLoader} */
	this._loader = new THREE.ImageLoader();

	/** @type {ImageLoadResult} */
	this._result = null;
};

/**
 * @return {ImageLoadResult}
 */
CachedImageLoader.prototype.getResult = function () {
	return this._result;
};

/**
 * @param {string} url
 * @param callback
 */
CachedImageLoader.prototype.loadImage = function (url, callback) {
	var self = this;
	this._loader.load(url, function (image) {
		// done
		var result = new ImageLoadResult();
		result.url = url;
		result.success = true;
		result.image = image;
		self._result = result;
		if (callback) {
			callback();
		}
	}, undefined, function (event) // eslint-disable-line no-unused-vars
	{
		// error
		var result = new ImageLoadResult();
		result.url = url;
		result.success = false;
		self._result = result;
		if (callback) {
			callback();
		}
	});
};

module.exports = CachedImageLoader;
