// animation-utilities/lib/debug/index.js

"use strict";

if (global._animationutilities_singleton) {
	module.exports = global._animationutilities_singleton;
} else {

	/** @type {JiboConfig} */
	module.exports.JiboConfig = require("./geometry-info/JiboConfig");

	/** @type {RobotInfo} */
	module.exports.RobotInfo = require("./geometry-info/RobotInfo");

	/** @type {RobotInfo} */
	module.exports.EyeKinematicsHelper = require("./geometry-info/EyeKinematicsHelper");

	/** @type {animate} */
	module.exports.animate = require("./animation-animate/AnimateImpl");

	/** @type {visualize} */
	module.exports.visualize = require("./animation-visualize/VisualizeImpl");

	/** @type {TimelineBuilder} */
	module.exports.TimelineBuilder = require("./animation-macros/TimelineBuilder");

	module.exports.Time = require("./ifr-core/Time");
	module.exports.Clock = require("./ifr-core/Clock");
	module.exports.slog = require("./ifr-core/SLog");

	module.exports.MotionInterface = require("./animation-body/MotionInterface");

	module.exports.body = {
		/** @type {BodyOutput} */
		BodyVelocityOutput: require("./animation-body/BodyVelocityOutput"),
		/** @type {BodyOutput} */
		BodyTrajectoryOutput: require("./animation-body/BodyTrajectoryOutput"),
		/** @type {BodyOutput} */
		BodyPositionOutput: require("./animation-body/BodyPositionOutput"),
		/** @type {BodyOutput} */
		BodyPosVelOutput: require("./animation-body/BodyPosVelOutput"),
		/** @type {BodyOutput} */
		BodyPosVelComboOutput: require("./animation-body/BodyPosVelComboOutput"),
		/** @type {BodyOutput} */
		MotionServiceOutput: require("./animation-body/MotionServiceOutput"),
		/** @type {MotionLog} */
		MotionLog: require("./animation-body/MotionLog")
	};

	module.exports.ui = {
		/** @type {Bakery} */
		Bakery: require("./ifr-core/Bakery"),
		/** @type {JSONBaker} */
		JSONBaker: require("./ifr-core/JSONBaker")
	};

	//base classes used for defining client subclasses
	module.exports.protos = {
		/** @type {SampleCombiner} */
		SampleCombiner: require("./animation-animate/timeline/SampleCombiner"),
		/** @type {RenderPlugin} */
		RenderPlugin: require("./animation-visualize/RenderPlugin")
	};

	//classes to help with graphical display/interfaces
	module.exports.graphics = {
		/** @type {AnchoredTargetVisualizer} */
		AnchoredTargetVisualizer: require("./ifr-visualizer/AnchoredTargetVisualizer"),
		/** @type {GLTextOverlayPool} */
		GLTextOverlayPool: require("./ifr-visualizer/GLTextOverlayPool"),
		/** @type {GLLinePool} */
		GLLinePool: require("./ifr-visualizer/GLLinePool")
	};

	module.exports.LEDOutput = require("./animation-body/LEDOutput");
	module.exports.AuxOutput = require("./animation-animate/timeline/AuxOutput");

	module.exports.MouseCoordinateWrangler = require("./ifr-visualizer/MouseCoordinateWrangler");
	module.exports.MouseTargetPositioner = require("./ifr-visualizer/MouseTargetPositioner");

	module.exports.TrajectoryControllerSim = require("./ifr-motion/feedback/TrajectoryControllerSim");
	module.exports.PosVelControllerSim = require("./ifr-motion/feedback/PosVelControllerSim");

	//we must use the same three instance when adding geometry
	module.exports.THREE = require('three');

	//initialize default logging channels
	module.exports.slog.setPrintChannels(["ERROR",
	//"WARN",
	//"INFO",
	"ACCEL_PLANNER", "CALIBRATION",
	//"LOOKAT",
	"MODEL_LOADING",
	//"MOUSE_COORD_WRANGLER",
	"RENDER_PLUGIN"
	//"UI_TARGET"
	]);

	global._animationutilities_singleton = module.exports;
}
