// jibo/lib/debug/rendering/Eye.js
"use strict";
var __extends = (this && this.__extends) || function (d, b) {
    for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p];
    function __() { this.constructor = d; }
    d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
};
var AbstractEye_1 = require('./AbstractEye');
var jibo_keyframes_1 = require('jibo-keyframes');
var Eye = (function (_super) {
    __extends(Eye, _super);
    function Eye() {
        _super.apply(this, arguments);
    }
    Eye.prototype.update = function (timestamp, dofValues) {
        if (!this.connected) {
            return;
        }
        var vertices = this.eyeMesh.dofs;
        if (dofValues.eyeTextureInfixBn_r !== this.texturePrev) {
            this.texturePrev = dofValues.eyeTextureInfixBn_r;
            this.texture = this.texturePrev;
        }
        vertices[0].x = jibo_keyframes_1.conversion.toPixelsX(dofValues.vertexJoint1_t);
        vertices[0].y = jibo_keyframes_1.conversion.toPixelsY(dofValues.vertexJoint1_t_2);
        vertices[1].x = jibo_keyframes_1.conversion.toPixelsX(dofValues.vertexJoint2_t);
        vertices[1].y = jibo_keyframes_1.conversion.toPixelsY(dofValues.vertexJoint2_t_2);
        vertices[2].x = jibo_keyframes_1.conversion.toPixelsX(dofValues.vertexJoint3_t);
        vertices[2].y = jibo_keyframes_1.conversion.toPixelsY(dofValues.vertexJoint3_t_2);
        vertices[3].x = jibo_keyframes_1.conversion.toPixelsX(dofValues.vertexJoint4_t);
        vertices[3].y = jibo_keyframes_1.conversion.toPixelsY(dofValues.vertexJoint4_t_2);
        vertices[4].x = jibo_keyframes_1.conversion.toPixelsX(dofValues.vertexJoint5_t);
        vertices[4].y = jibo_keyframes_1.conversion.toPixelsY(dofValues.vertexJoint5_t_2);
        vertices[5].x = jibo_keyframes_1.conversion.toPixelsX(dofValues.vertexJoint6_t);
        vertices[5].y = jibo_keyframes_1.conversion.toPixelsY(dofValues.vertexJoint6_t_2);
        vertices[6].x = jibo_keyframes_1.conversion.toPixelsX(dofValues.vertexJoint7_t);
        vertices[6].y = jibo_keyframes_1.conversion.toPixelsY(dofValues.vertexJoint7_t_2);
        vertices[7].x = jibo_keyframes_1.conversion.toPixelsX(dofValues.vertexJoint8_t);
        vertices[7].y = jibo_keyframes_1.conversion.toPixelsY(dofValues.vertexJoint8_t_2);
        vertices[8].x = jibo_keyframes_1.conversion.toPixelsX(dofValues.vertexJoint9_t);
        vertices[8].y = jibo_keyframes_1.conversion.toPixelsY(dofValues.vertexJoint9_t_2);
        this.x = jibo_keyframes_1.conversion.toPixelsX(dofValues.eyeSubRootBn_t);
        this.y = -jibo_keyframes_1.conversion.toPixelsX(dofValues.eyeSubRootBn_t_2);
        this.red = dofValues.eye_redChannelBn_r;
        this.green = dofValues.eye_greenChannelBn_r;
        this.blue = dofValues.eye_blueChannelBn_r;
        this.rotation = -dofValues.eyeSubRootBn_r;
    };
    return Eye;
}(AbstractEye_1.default));
Object.defineProperty(exports, "__esModule", { value: true });
exports.default = Eye;
