// jibo/lib/debug/rendering/EyeMesh.js
"use strict";
var __extends = (this && this.__extends) || function (d, b) {
    for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p];
    function __() { this.constructor = d; }
    d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
};
var DOF_1 = require("./DOF");
var EyeMesh = (function (_super) {
    __extends(EyeMesh, _super);
    function EyeMesh(t) {
        _super.call(this, t, null, null, null, PIXI.mesh.Mesh.DRAW_MODES.TRIANGLES);
        this.uvs = new Float32Array([
            0, 0,
            0.5, 0,
            1, 0,
            0, 0.5,
            0.5, 0.5,
            1, 0.5,
            0, 1,
            0.5, 1,
            1, 1
        ]);
        this.vertices = new Float32Array(this.uvs.length * 2);
        this.dofs = [];
        for (var i = 0; i < this.uvs.length; i += 2) {
            this.dofs.push(new DOF_1.default(this.uvs[i], this.uvs[i + 1]));
        }
        this.indices = new Uint16Array([
            0, 1, 3,
            1, 3, 4,
            1, 4, 5,
            2, 5, 1,
            3, 4, 7,
            7, 6, 3,
            4, 5, 7,
            5, 7, 8
        ]);
    }
    EyeMesh.prototype.updateTransform = function () {
        for (var i = 0; i < this.dofs.length; i++) {
            var point = this.dofs[i];
            var index = i * 2;
            this.vertices[index] = point.x;
            this.vertices[index + 1] = point.y;
        }
        this.containerUpdateTransform();
    };
    return EyeMesh;
}(PIXI.mesh.Mesh));
Object.defineProperty(exports, "__esModule", { value: true });
exports.default = EyeMesh;
