// jibo/lib/debug/rendering/EyeOverlay.js
"use strict";
var __extends = (this && this.__extends) || function (d, b) {
    for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p];
    function __() { this.constructor = d; }
    d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
};
var AbstractEye_1 = require('./AbstractEye');
var jibo_keyframes_1 = require('jibo-keyframes');
var EyeOverlay = (function (_super) {
    __extends(EyeOverlay, _super);
    function EyeOverlay() {
        _super.apply(this, arguments);
    }
    EyeOverlay.prototype.update = function (timestamp, dofValues) {
        if (!this.connected) {
            return;
        }
        var vertices = this.eyeMesh.dofs;
        if (dofValues.eyeTextureInfixBn_r !== this.texturePrev) {
            this.texturePrev = dofValues.overlayTextureInfixBn_r;
            this.texture = this.texturePrev;
        }
        vertices[0].x = jibo_keyframes_1.conversion.toPixelsX(dofValues.overlay_vertexJoint1_t);
        vertices[0].y = jibo_keyframes_1.conversion.toPixelsY(dofValues.overlay_vertexJoint1_t_2);
        vertices[1].x = jibo_keyframes_1.conversion.toPixelsX(dofValues.overlay_vertexJoint2_t);
        vertices[1].y = jibo_keyframes_1.conversion.toPixelsY(dofValues.overlay_vertexJoint2_t_2);
        vertices[2].x = jibo_keyframes_1.conversion.toPixelsX(dofValues.overlay_vertexJoint3_t);
        vertices[2].y = jibo_keyframes_1.conversion.toPixelsY(dofValues.overlay_vertexJoint3_t_2);
        vertices[3].x = jibo_keyframes_1.conversion.toPixelsX(dofValues.overlay_vertexJoint4_t);
        vertices[3].y = jibo_keyframes_1.conversion.toPixelsY(dofValues.overlay_vertexJoint4_t_2);
        vertices[4].x = jibo_keyframes_1.conversion.toPixelsX(dofValues.overlay_vertexJoint5_t);
        vertices[4].y = jibo_keyframes_1.conversion.toPixelsY(dofValues.overlay_vertexJoint5_t_2);
        vertices[5].x = jibo_keyframes_1.conversion.toPixelsX(dofValues.overlay_vertexJoint6_t);
        vertices[5].y = jibo_keyframes_1.conversion.toPixelsY(dofValues.overlay_vertexJoint6_t_2);
        vertices[6].x = jibo_keyframes_1.conversion.toPixelsX(dofValues.overlay_vertexJoint7_t);
        vertices[6].y = jibo_keyframes_1.conversion.toPixelsY(dofValues.overlay_vertexJoint7_t_2);
        vertices[7].x = jibo_keyframes_1.conversion.toPixelsX(dofValues.overlay_vertexJoint8_t);
        vertices[7].y = jibo_keyframes_1.conversion.toPixelsY(dofValues.overlay_vertexJoint8_t_2);
        vertices[8].x = jibo_keyframes_1.conversion.toPixelsX(dofValues.overlay_vertexJoint9_t);
        vertices[8].y = jibo_keyframes_1.conversion.toPixelsY(dofValues.overlay_vertexJoint9_t_2);
        this.x = jibo_keyframes_1.conversion.toPixelsX(dofValues.overlay_textureSubRootBn_t);
        this.y = -jibo_keyframes_1.conversion.toPixelsX(dofValues.overlay_textureSubRootBn_t_2);
        this.red = dofValues.overlay_redChannelBn_r;
        this.green = dofValues.overlay_greenChannelBn_r;
        this.blue = dofValues.overlay_blueChannelBn_r;
        this.rotation = -dofValues.overlay_textureSubRootBn_r;
    };
    return EyeOverlay;
}(AbstractEye_1.default));
Object.defineProperty(exports, "__esModule", { value: true });
exports.default = EyeOverlay;
